File: leetcode/2536-increment-submatrices-by-one/Solution.java
import java.util.Arrays;

/**
 * LeetCode 2536. Increment Submatrices by One
 * 难度：Medium
 *
 * 解题思路（二维差分数组）：
 * 朴素做法对每个查询遍历子矩阵是 O(n²*q)。这里用二维差分数组优化到 O(n² + q)：
 *   1. 创建 (n+1) x (n+1) 的差分数组 diff。
 *   2. 对每个查询 [r1, c1, r2, c2]，在差分数组上四个点打标记:
 *        diff[r1][c1]     += 1
 *        diff[r1][c2+1]   -= 1   // 注意边界：c2+1 < n+1
 *        diff[r2+1][c1]   -= 1   // 注意边界：r2+1 < n+1
 *        diff[r2+1][c2+1] += 1   // 注意边界
 *   3. 二维前缀和还原：对每一行先做一次前缀和，再对每一列做一次前缀和。
 *
 * 时间复杂度：O(n² + q)
 * 空间复杂度：O(n²)
 */
public class Solution {
    public int[][] rangeAddQueries(int n, int[][] queries) {
        // 创建 (n+1) x (n+1) 的差分数组
        // 多出一行一列的目的：当 r2+1 == n 或 c2+1 == n 时不会越界
        int[][] diff = new int[n + 1][n + 1];

        // 对每个查询在差分数组上"打标记"
        for (int[] q : queries) {
            int r1 = q[0], c1 = q[1], r2 = q[2], c2 = q[3];
            diff[r1][c1] += 1;
            diff[r1][c2 + 1] -= 1;       // 左闭右开区间标记上界
            diff[r2 + 1][c1] -= 1;       // 左闭右开区间标记下界
            diff[r2 + 1][c2 + 1] += 1;   // 右下角的补偿项
        }

        // 先做行方向的前缀和（即对每一行从左到右累加）
        for (int i = 0; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                diff[i][j] += diff[i][j - 1];
            }
        }

        // 再做列方向的前缀和（即对每一列从上到下累加）
        for (int j = 0; j <= n; j++) {
            for (int i = 1; i <= n; i++) {
                diff[i][j] += diff[i - 1][j];
            }
        }

        // 取前 n x n 部分即为最终矩阵
        int[][] ans = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                ans[i][j] = diff[i][j];
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // 测试用例 1
        int n1 = 3;
        int[][] q1 = {{1, 1, 2, 2}, {0, 0, 1, 1}};
        int[][] r1 = sol.rangeAddQueries(n1, q1);
        int[][] e1 = {{1, 1, 0}, {1, 2, 1}, {0, 1, 1}};
        System.out.println("Test 1: " + Arrays.deepEquals(r1, e1) + ", result=" + Arrays.deepToString(r1));

        // 测试用例 2
        int n2 = 2;
        int[][] q2 = {{0, 0, 1, 1}};
        int[][] r2 = sol.rangeAddQueries(n2, q2);
        int[][] e2 = {{1, 1}, {1, 1}};
        System.out.println("Test 2: " + Arrays.deepEquals(r2, e2) + ", result=" + Arrays.deepToString(r2));

        // 测试用例 3：整个矩阵被增量两次
        int n3 = 3;
        int[][] q3 = {{0, 0, 2, 2}, {0, 0, 2, 2}};
        int[][] r3 = sol.rangeAddQueries(n3, q3);
        int[][] e3 = {{2, 2, 2}, {2, 2, 2}, {2, 2, 2}};
        System.out.println("Test 3: " + Arrays.deepEquals(r3, e3) + ", result=" + Arrays.deepToString(r3));

        // 测试用例 4：n=1 单点查询
        int n4 = 1;
        int[][] q4 = {{0, 0, 0, 0}};
        int[][] r4 = sol.rangeAddQueries(n4, q4);
        int[][] e4 = {{1}};
        System.out.println("Test 4: " + Arrays.deepEquals(r4, e4) + ", result=" + Arrays.deepToString(r4));
    }
}

File: leetcode/2536-increment-submatrices-by-one/solution.py
"""
LeetCode 2536. Increment Submatrices by One
难度：Medium

题目描述：
给定一个 n x n 的矩阵 mat，初始全为 0。对每个查询 [row1, col1, row2, col2]，
将子矩阵 (row1, col1) 到 (row2, col2) 中所有元素加 1。返回处理完所有查询后的矩阵。

解题思路（二维差分数组）：
朴素做法是对每个查询遍历整个子矩阵，时间复杂度 O(n² * q)，最坏情况
n=500, q=10⁴ 时为 2.5*10⁹，会超时。
利用二维差分数组可以做到 O(n² + q)：
  1. 创建一个 (n+1) x (n+1) 的差分数组 diff，初始全 0。
  2. 对每个查询 [r1, c1, r2, c2]，执行四个"打标记"操作：
        diff[r1][c1]     += 1
        diff[r1][c2 + 1] -= 1   (若 c2 + 1 < n)
        diff[r2 + 1][c1] -= 1   (若 r2 + 1 < n)
        diff[r2 + 1][c2 + 1] += 1 (若 r2 + 1 < n 且 c2 + 1 < n)
  3. 对 diff 做二维前缀和还原，得到每个位置的最终值。

时间复杂度：O(n² + q)
空间复杂度：O(n²)
"""


class Solution:
    def rangeAddQueries(self, n: int, queries: list[list[int]]) -> list[list[int]]:
        # 创建 (n+1) x (n+1) 的差分数组，多出一行一列用于处理边界
        diff = [[0] * (n + 1) for _ in range(n + 1)]

        # 对每个查询在差分数组上"打标记"
        for r1, c1, r2, c2 in queries:
            diff[r1][c1] += 1
            diff[r1][c2 + 1] -= 1
            diff[r2 + 1][c1] -= 1
            diff[r2 + 1][c2 + 1] += 1

        # 二维前缀和还原：第一遍沿列方向累加，第二遍沿行方向累加
        # 先做行方向的前缀和
        for i in range(n + 1):
            for j in range(1, n + 1):
                diff[i][j] += diff[i][j - 1]

        # 再做列方向的前缀和（即把每一列向下累加）
        for j in range(n + 1):
            for i in range(1, n + 1):
                diff[i][j] += diff[i - 1][j]

        # 取前 n x n 部分即为答案
        return [row[:n] for row in diff[:n]]


if __name__ == "__main__":
    # 测试用例 1
    sol = Solution()
    result1 = sol.rangeAddQueries(3, [[1, 1, 2, 2], [0, 0, 1, 1]])
    expected1 = [[1, 1, 0], [1, 2, 1], [0, 1, 1]]
    print(f"Test 1: {result1 == expected1}, result={result1}")

    # 测试用例 2
    result2 = sol.rangeAddQueries(2, [[0, 0, 1, 1]])
    expected2 = [[1, 1], [1, 1]]
    print(f"Test 2: {result2 == expected2}, result={result2}")

    # 测试用例 3：没有查询（不合法，但测一下 n=1 的单点情况）
    result3 = sol.rangeAddQueries(1, [[0, 0, 0, 0]])
    expected3 = [[1]]
    print(f"Test 3: {result3 == expected3}, result={result3}")

    # 测试用例 4：覆盖整个矩阵
    result4 = sol.rangeAddQueries(3, [[0, 0, 2, 2], [0, 0, 2, 2]])
    expected4 = [[2, 2, 2], [2, 2, 2], [2, 2, 2]]
    print(f"Test 4: {result4 == expected4}, result={result4}")
